File: csharp/two-bucket/TwoBucket.cs
public enum Bucket
{
    One,
    Two
}

public class TwoBucketResult
{
    public int Moves { get; set; }
    public Bucket GoalBucket { get; set; }
    public int OtherBucket { get; set; }
}

public class TwoBucket(int bucketOne, int bucketTwo, Bucket startBucket)
{
    private readonly int bucketOne = bucketOne;
    private readonly int bucketTwo = bucketTwo;
    private readonly Bucket startBucket = startBucket;

    public TwoBucketResult Measure(int goal)
    {
        if (goal > Math.Max(bucketOne, bucketTwo) || goal % Gcd(bucketOne, bucketTwo) != 0) throw new ArgumentException();

        var start = startBucket == Bucket.One ? (bucketOne, 0) : (0, bucketTwo);
        var queue = new Queue<((int b1, int b2) state, int moves)>();
        var visited = new HashSet<(int, int)>();

        queue.Enqueue((start, 1));
        visited.Add(start);

        while (queue.Count > 0)
        {
            var ((b1, b2), moves) = queue.Dequeue();

            if (b1 == goal || b2 == goal)
            {
                return new TwoBucketResult
                {
                    Moves = moves,
                    GoalBucket = b1 == goal ? Bucket.One : Bucket.Two,
                    OtherBucket = b1 == goal ? b2 : b1,
                };
            }

            var nextStates = new List<(int, int)>
            {
                (bucketOne, b2),
                (b1, bucketTwo),
                (0, b2),
                (b1, 0),  
            };

            var amount = Math.Min(b1, bucketTwo - b2);
            nextStates.Add((b1 - amount, b2 + amount));

            amount = Math.Min(b2, bucketOne - b1);
            nextStates.Add((b1 + amount, b2 - amount));

            foreach(var state in nextStates)
            {
                if (startBucket == Bucket.One && state == (0, bucketTwo)) continue;

                if (startBucket == Bucket.Two && state == (bucketOne, 0)) continue;

                if (visited.Add(state))
                {
                    queue.Enqueue((state, moves + 1));
                }
            }
        }

        throw new ArgumentException();
    }

    private static int Gcd(int a, int b)
    {
        while (b != 0)
        {
            (a, b) = (b, a % b);
        }

        return a;
    }
}

File: cpp/two-bucket/two_bucket.cpp
#include "two_bucket.h"
#include <algorithm>
#include <stdexcept>
#include <queue>
#include <set>

namespace two_bucket
{

    static int gcd(int a, int b)
    {
        while (b != 0)
        {
            int temp = b;
            b = a % b;
            a = temp;
        }

        return a;
    }

    measure_result measure(int bucket1_capacity, int bucket2_capacity, int target_volume, bucket_id start_bucket)
    {
        if (target_volume > std::max(bucket1_capacity, bucket2_capacity) ||
            target_volume % gcd(bucket1_capacity, bucket2_capacity) != 0)
        {
            throw std::domain_error("Impossible");
        }

        std::pair<int, int> start =
            (start_bucket == bucket_id::one)
                ? std::make_pair(bucket1_capacity, 0)
                : std::make_pair(0, bucket2_capacity);

        std::queue<std::pair<std::pair<int, int>, int>> q;
        std::set<std::pair<int, int>> visited;

        q.push({start, 1});
        visited.insert(start);

        while (!q.empty())
        {
            auto current = q.front();
            q.pop();

            int b1 = current.first.first;
            int b2 = current.first.second;
            int moves = current.second;

            if (b1 == target_volume || b2 == target_volume)
            {
                measure_result result;
                result.num_moves = moves;
                result.goal_bucket =
                    (b1 == target_volume) ? bucket_id::one : bucket_id::two;
                result.other_bucket_volume =
                    (b1 == target_volume) ? b2 : b1;
                return result;
            }

            std::vector<std::pair<int, int>> next_states = {
                {bucket1_capacity, b2},
                {b1, bucket2_capacity},
                {0, b2},
                {b1, 0}};

            int amount = std::min(b1, bucket2_capacity - b2);
            next_states.push_back({b1 - amount, b2 + amount});

            amount = std::min(b2, bucket1_capacity - b1);
            next_states.push_back({b1 + amount, b2 - amount});

            for (const auto &state : next_states)
            {
                if (start_bucket == bucket_id::one &&
                    state == std::make_pair(0, bucket2_capacity))
                {
                    continue;
                }

                if (start_bucket == bucket_id::two &&
                    state == std::make_pair(bucket1_capacity, 0))
                {
                    continue;
                }

                if (visited.insert(state).second)
                {
                    q.push({state, moves + 1});
                }
            }
        }

        throw std::domain_error("Impossible");
    }

} // namespace two_bucket
